library/core/src/intrinsics/mod.rs

```diff
@@ -4537,6 +4537,32 @@ pub(crate) const fn miri_promise_symbolic_alignment(ptr: *const (), align: usize
     )
 }
 
+#[requires(crate::mem::size_of::<T>() >= crate::mem::size_of::<U>())] //U cannot be larger than T
```

Comment:
We don't need to size requirement since this is a compilation error according to the `transmute` documentation. We do need to state that input `T` is a valid `U` value. One possibility would be to use the `can_dereference` method from `ub_checks`.

However, with Kani, the value validity checks are still incomplete (see https://github.com/model-checking/kani/issues/3203) for current limitations.

Reply:
I created some new commits with these suggestions in mind. The only thing I left unchanged for now is the check for equality in the sizes of the input and output, since according to the [documentation](https://doc.rust-lang.org/std/intrinsics/fn.transmute_unchecked.html) of `transmute_unchecked`, it seems that the compiler allows the inputs and outputs to have different sizes (as opposed to `transmute`), but considers this to be UB.